464 - Simplify Signup Flow

timbot1789 opened the pull request:
## This PR:
Resolves #464 
 
**1.** Rebuilds the user registration logic using community solid server V7's [automation API](https://communitysolidserver.github.io/CommunitySolidServer/latest/usage/account/json-api/#controlsaccountclientcredentials). This allows us to build out a user's entire pod with no action needed from the user other than username and password. This system only works on Community Solid Server V7, but that has been the case.
**2.** Adds a couple more states to the registration page. Now the page can show: creation form, loading page, and finish page. A future PR will need to add an error page for if there's an error in pod creation
**3.** Added a signup button to the home screen that takes a new user to the signup process. This won't set them up with a case manager, but it will make new volunteer onboarding significantly easier.
**4.** Added some code to outline the IDP logout feature that comes with inrupt's V2 client auth libraries (allows logout from both pod and app in one click). However, this is proving complicated, especially to deploy, so finishing the feature will be left to another PR. I don't want the code to sit on my computer forever, though.

## Future Steps/PRs Needed to Finish This Work (optional):

We will need to improve the error handling of the signup flow in a future PR. This PR only implements the happy path. It doesn't handle any errors during pod creation.

timbot1789 committed:
Add smoe stuff

timbot1789 committed:
update inrupt libraries to v2

timbot1789 committed:
Update session context to new events api

timbot1789 committed:
Partial implementation

timbot1789 committed:
Restore pod creation in CSS V7

timbot1789 committed:
Implement pod creation

timbot1789 committed:
Restore invite link

timbot1789 committed:
Create contact list on user creation

timbot1789 committed:
Add signup button

timbot1789 committed:
Fix tests

timbot1789 commented on config/solid-server/solid-config.json:
```diff
@@ -1,35 +1,37 @@
-    "css:CommunitySolidServer/config/util/representation-conversion/default.json",
-    "css:CommunitySolidServer/config/util/resource-locker/file.json",
-    "css:CommunitySolidServer/config/util/variables/default.json"
+    "css:config/app/init/static-root.json",
```
Created using new CSS  config generator:
https://communitysolidserver.github.io/configuration-generator/v7/

timbot1789 commented on env.template:
```diff
@@ -1,3 +1,4 @@
 VITE_SOLID_IDENTITY_PROVIDER="http://localhost:3000/"
 VITE_SOLID_POD_SERVER="http://localhost:3000/"
 VITE_SUGGESTED_OIDC_OPTIONS="http://localhost:3000/, https://opencommons.net/, https://solidcommunity.net/, https://login.inrupt.com/, https://inrupt.net/"
+VITE_CLIENT_ID_DOC="http://localhost:3000/clientAppId.json"
```
The client ID is necessary for allowing proper redirects when logging in or logging out. However, it must be hosted on the internet to properly use. This hosting issue won't be resolved here.

timbot1789 commented on package.json:
```diff
@@ -95,7 +96,7 @@
   },
   "lint-staged": {
-    "*.{js,css,ts,tsx,jsx}": [
+    "src/**/*.{js,css,ts,tsx,jsx}": [
```
Fix for vite config getting linted and causing the commit to fail.

timbot1789 commented on public/clientAppId.jsonld:
```diff
@@ -0,0 +1,11 @@
+{
```
See [here](https://docs.inrupt.com/developer-tools/javascript/client-libraries/tutorial/authenticate-client/) for more information on this document

timbot1789 commented on src/AppRoutes.jsx:
```diff
@@ -32,14 +31,7 @@ const AppRoutes = () => {
-          </SessionProvider>
-        }
-      />
+      <Route path="/signup" element={<Signup />} />
```
The signup now manages its own state entirely (it doesn't interact with the session) so we don't need to mask the session or force the user to log out.

timbot1789 commented on src/components/Signup/podSignupHelpers/initializePod.js:
```diff
@@ -0,0 +1,90 @@
+// Inrupt Imports
```
This was incorrectly labeled as a `jsx` file. It's a pure JS file.

andycwilliams replied:
Is there a functional difference in this case, or is it more a matter of best practices?

timbot1789 replied:
Best practice. Also makes dev tools understand it better.

timbot1789 commented on src/components/Signup/podSignupHelpers/registerPod.js:
```diff
@@ -0,0 +1,102 @@
+import {
```
This was incorrectly labeled as a jsx file. It's a pure js style

timbot1789 commented on src/contexts/SessionContext.jsx:
```diff
@@ -57,7 +57,7 @@ export const SessionProvider = ({
   useEffect(() => {
     if (onSessionRestore !== undefined) {
-      onSessionRestoreClient(onSessionRestore);
+      session.events.on(EVENTS.SESSION_RESTORED, onSessionRestore);
```
Fixes for Inrupt's v2 library upgrade. `onSessionRestoreClient` and all other `onEvent` functions have been removed and replaced with this `session.events.on` style.

timbot1789 commented on vite.config.js:
```diff
@@ -2,40 +2,34 @@ import { defineConfig } from 'vite';
-// https://github.com/inrupt/solid-client-authn-js/issues/3001
-// Once that issue is fixed and the solid-client-authn-js version is bumped
-// We can delete this function and all related code
-function withCustomResolver(
```
Inrupt's V2 libraries now support ES Modules out of the box, so this custom resolver is no longer needed.

andycwilliams commented on src/components/Signup/ShowNewPod.jsx:
```diff
@@ -32,6 +34,10 @@ const ShowNewPod = ({ oidcIssuer }) => {
           ? `You have registered with ${data[0].person}`
           : 'You have not registered with a case manager'}
       </Typography>
+      <Typography>Click here to return to the home page and log into PASS:</Typography>
```
Maybe upon successful registration it could countdown and then redirect automatically? I've seen plenty of instances along those lines (e.g. "[action successful], you will be redirected in [X] seconds or click here")

timbot1789 replied:
I agree we can definitely improve this page. It's a placeholder right now before we settle on a real design. A timed countdown is one approach.

I wrote up an issue here related to it:
https://github.com/codeforpdx/PASS/issues/441

One of the ideas I had was to include an overview of what the pod is, and what kind of things it lets you do. This wouldn't work with the timer idea, since the overview would take a lot of time to read, and users may want to be linked to other resources to read more.

Either way it's not related to this PR. We can continue the conversation on that issue.

andycwilliams commented on src/pages/Signup.jsx:
```diff
@@ -29,28 +27,34 @@ import { PodRegistrationForm, ShowNewPod, initializePod, registerPod } from '@co
+    setStep('done');
   };
 
   const loadProfileInfo = async () => {
```
Receiving this error when first clicking on Sign Up

<img width="217" alt="2024-02-25 (2)" src="https://github.com/codeforpdx/PASS/assets/83156697/87ebd8e4-08ac-444a-9fca-29c475075b3a">

timbot1789 replied:
This should have only happened when clicking the signup button. Not when using the invite link.

This is now fixed.

andycwilliams reviewed (commented):
Haven't had a chance to go through all the changes. These are just immediate thoughts.

Slightly tangential: I'm thinking about contributing to #547, which I believe affects the same `Signup` component. Thus it might be an opportunity to reassess that and its related issue (#463), as the needs may have changed since then.

It may also be a good idea to add a Cancel/Return button on the Sign Up page. Or perhaps that could be part of the above PR instead.

leekahung commented:
I've cleared out my local_temp_server_files directory to test out the new build before diving into the codebase changes. The new signup functionality seems to be running fine, I see the new pods being created from the process. One small issue I've saw pop-up was the webId not being shown (see screenshot).

![Screenshot 2024-02-25 at 4 37 17 PM](https://github.com/codeforpdx/PASS/assets/14917816/3c5946b1-0533-4937-91a2-87765f8e8c6c)

I've made a few more accounts to test things out, but it looks like the problem above was related to who logged in last as it show the previous logged in user's webId (see clips).

(Previous user logged in was `http://localhost:3000/newtestuser/profile/card#me`)

https://github.com/codeforpdx/PASS/assets/14917816/b36ba706-c738-4f1b-ade1-80c301949ce7

(Previous user logged in was `http://localhost:3000/testpod/profile/card#me`)

https://github.com/codeforpdx/PASS/assets/14917816/ff3aac94-18b6-4503-adcb-40b5892b1c93

If the server is fresh and no one had logged in last, it makes sense that the webId was blank for the first pod creation, but it should have been the webId of the user creating the webId, not the previous logged in user.

Another thing I've noticed when signing up using that button was that the .acl file for `/PASS/Documents` failed to generate for some of the new users. It returns a 409 error instead when logging in for the first time. Without it, we can't control the permissions for "Documents".

![Screenshot 2024-02-25 at 4 39 04 PM](https://github.com/codeforpdx/PASS/assets/14917816/d9a55920-8c2e-405f-ac15-e93bfbd59f15)

If we were to re-login (logout and login again), the .acl file gets generated as it should if the file was missing (see screenshot), but the function for generating the .acl file should have been running correctly the first time around, so I'm not sure what happened there.

![Screenshot 2024-02-25 at 4 43 28 PM](https://github.com/codeforpdx/PASS/assets/14917816/daebcf95-c7cd-42e9-a0bc-12f5b9e609af)

If I were to signup for a pod using the signup forms by CSS instead, the 409 error doesn't seem to appear at all (see screenshot). The .acl file gets generated the first time around.
 
![Screenshot 2024-02-25 at 4 45 41 PM](https://github.com/codeforpdx/PASS/assets/14917816/03394903-6343-4ed7-99c7-e535fc3ccc40)

Something not related to this PR (but an earlier one), is that user deletion fails to trigger when the first and last name are null. I'll be posting a separate bug report for that (see issue #585).

Other than the issues I've seen above, all other functionalities seem to be work correctly like document uploads/preview/deletion, adding new contacts, sending messages, etc. Sharing documents still work as long as the .acl file exists within `/PASS/Documents/` inside the user's pod

timbot1789 committed:
Fix signup error when no case manager

timbot1789 committed:
Fix sign up finish displaying wrong values

timbot1789 committed:
Create doc list with ACL on pod creation

timbot1789 commented:
@leekahung Thanks for the thorough testing. Fixed the web ID display and the pod URL display (it was broken too).

Added the doclist to pod initialization. It's an empty document. The ACL doesn't need to be created at this time. The doclist inherits permissions from the PASS pod, and the document sharing feature creates an acl if one doesn't already exist.

leekahung commented:
> @leekahung Thanks for the thorough testing. Fixed the web ID display and the pod URL display (it was broken too).
> 
> Added the doclist to pod initialization. It's an empty document. The ACL doesn't need to be created at this time. The doclist inherits permissions from the PASS pod, and the document sharing feature creates an acl if one doesn't already exist.

All good, I'll give it another go through post changes.

leekahung commented:
Alright, just gone through the changes. The previous errors I was getting before for the features are now gone. So things are looking good there. I'll take a look over the code.

leekahung commented on src/components/Signup/ShowNewPod.jsx:
```diff
@@ -5,33 +5,38 @@ import React from 'react';
 // Hooks imports
-import useSession from '@hooks/useSession';
 import { useContactsList } from '@hooks';
+import { Button } from '@mui/material';
```
Very minor, recommend import style to be `import Button from '@mui/material/Button'` to be consistent with other MUI imports

leekahung commented on src/pages/Signup.jsx:
```diff
@@ -11,14 +11,12 @@ import Box from '@mui/material/Box';
 
 // Signup Form Imports
 import { PodRegistrationForm, ShowNewPod, initializePod, registerPod } from '@components/Signup';
+import { Typography } from '@mui/material';
```
Very minor nitpick similar to the previous comment about MUI imports and consistency.

leekahung reviewed (approved):
Outside of two minor syntax comments, the rest of the code looks good to me!

It's not too important now for development, but for the sign-up process, I was wondering if we should give a disclaimer somewhere pointing out when this feature could be used since it's limited to just servers with CSS v7 vs. when users should just use the login button to get to the sign-up page for their respective servers like with solidcommunity.net's server or Inrupt's PodSpaces to make their Pods.

![Screenshot 2024-02-25 at 11 50 41 PM](https://github.com/codeforpdx/PASS/assets/14917816/51ac0c37-67b7-4a78-9c52-23a1e7265b1e)
<img width="365" alt="Screenshot 2024-02-25 at 11 50 01 PM" src="https://github.com/codeforpdx/PASS/assets/14917816/3936c444-bb6a-408f-9ad0-1fb78cc5def7">

timbot1789 committed:
Show pod provider after signup

timbot1789 commented:
> Outside of two minor syntax comments, the rest of the code looks good to me!
> 
> It's not too important now for development, but for the sign-up process, I was wondering if we should give a disclaimer somewhere pointing out when this feature could be used since it's limited to just servers with CSS v7 vs. when users should just use the login button to get to the sign-up page for their respective servers like with solidcommunity.net's server or Inrupt's PodSpaces to make their Pods.
> 

With the current design, the user doesn't get to choose their pod provider during pod creation flow. They're required to use the pod provider configured in the running PASS instance. In that case, the disclaimer would need to be sysadmin/dev facing. "You can only connect this part of the app to a CSS v7 server." I've added disclaimers in the comments of the registration function code, which is sufficient unless we want to write a sysadmin guide in the wiki.

You're right though that we need to display the Pod Provider that PASS signed the user up with at the end. I added a line to the last page that shows the URL of the pod provider.

We can also add an option to let a user enroll with an existing pod. But that's a separate flow where we can expect the user to be able to handle pod creation without our help, so no disclaimer is needed. I just added [an issue](https://github.com/codeforpdx/PASS/issues/586) to the project for it. I had this feature in a demo a while back, but haven't written up tests for it yet.
